### PR TITLE
add methods for dealing with C++ namespaces

### PR DESCRIPTION
**Pull request by thewilsonator**

### CODE REVIEW

File: spec/cpp_interface.dd [hunk 1]
```diff
@@ -178,6 +178,71 @@ void main()
...
 ------
 
+    $(P C++ can open the same namespace in the same file and multiple files.
+      In D, this can be done as follows:)
+
+---
+module ns;
+extern (C++, `ns`) { int foo() { return 1; } }
+extern (C++, `ns`) { int bar() { return 2; } }
+---
```
[the remark this answers is not in the document]
Comment (thewilsonator) [reply spec/cpp_interface.dd:188]: I'll go with just exclusive the second example.

File: spec/cpp_interface.dd [hunk 2]
```diff
@@ -178,6 +178,71 @@ void main()
...
+      In D, this can be done as follows:)
+
+ ---
+ module ns;
+ extern (C++, `ns`) { int foo() { return 1; } }
+ extern (C++, `ns`) { int bar() { return 2; } }
+ ---
+ ---
+ import a;
+ static assert(ns.foo() == 1 && ns.bar() == 2);
```
[the remark this answers is not in the document]
Comment (thewilsonator) [reply spec/cpp_interface.dd]: Will go with the second one

File: spec/cpp_interface.dd [hunk 3]
```diff
@@ -178,6 +178,71 @@ void main()
...
+
+---
+module ns;
+extern (C++, `ns`) { int foo() { return 1; } }
+extern (C++, `ns`) { int bar() { return 2; } }
+---
+---
+import a;
+static assert(ns.foo() == 1 && ns.bar() == 2);
+---
```
[the remark this answers is not in the document]
Comment (thewilsonator) [reply spec/cpp_interface.dd:192]: It also affects the mangling

File: spec/cpp_interface.dd [hunk 4]
```diff
@@ -178,6 +178,76 @@ void main()
...
 
+    $(P C++ can open the same namespace in the same file and multiple files.
+      In D, this can be done as follows:)
+
+---
+module ns;
+extern (C++, `ns`) 
+{ 
+	int foo() { return 1; }
+    int bar() { return 2; }
```
Comment (PetarKirov) [inline spec/cpp_interface.dd:189]: weird indentation

File: spec/cpp_interface.dd [hunk 5]
```diff
@@ -178,6 +178,76 @@ void main()
...
+---
+module a; extern (C++, `ns`) { int foo() { return 1; } }
+---
+---
+module b; extern (C++, `ns`) { int bar() { return 2; } }
+---
+---
+module ns;
+public import a, b;
+---
```
Comment (PetarKirov) [inline spec/cpp_interface.dd:203]:
```suggestion
$(P or in multiple files, by organizing them in a package consisting of several modules:)
---
ns/
├── a.d
├── b.d
└── package.d
---

$(P File `ns/a.d`:)
---
module a; extern (C++, `ns`) { int foo() { return 1; } }
---

$(P File `ns/b.d`:)
---
module b; extern (C++, `ns`) { int bar() { return 2; } }
---

$(P File `ns/package.d`:)
---
module ns;
public import a, b;
---
```

File: spec/cpp_interface.dd [hunk 6]
```diff
@@ -178,6 +178,76 @@ void main()
...
+$(P Then you can import the module containing the extern C++ declarations as follows:)
```
Comment (PetarKirov) [inline spec/cpp_interface.dd:204]:
```suggestion
$(P Then you can import the package containing the extern C++ declarations as follows:)
```

File: spec/cpp_interface.dd [hunk 7]
```diff
@@ -178,6 +178,76 @@ void main()
...
+
+---
+import ns;
+static assert(foo() == 1 && bar() == 2);
+---
+
+$(P Note that the `extern (C++, `ns`)` linkage attribute affects only the ABI and mangling of
```
Comment (PetarKirov) [inline spec/cpp_interface.dd:211]: ABI includes name mangling:

```suggestion
$(P Note that the `extern (C++, `ns`)` linkage attribute affects only the ABI (name mangling and calling convention) of
```

https://en.wikipedia.org/wiki/Application_binary_interface#Complete_ABIs
> [..] ABIs standardize details such as the C++ name mangling,[2] exception propagation,[3] and calling convention between compilers on the same platform [..]

File: spec/cpp_interface.dd [hunk 8]
```diff
@@ -178,6 +178,76 @@ void main()
...
+---
+---
+module ns;
+import a, b;
+alias foo = a.ns.foo;
+alias bar = b.ns.bar;
+---
+---
+import ns;
+static assert(ns.foo() == 1 && ns.bar() == 2);
```
Comment (PetarKirov) [inline spec/cpp_interface.dd:248]:
```suggestion
static assert(foo() == 1 && bar() == 2);
```

Or use `static import`

File: spec/cpp_interface.dd (hunk 4 above)
Comment (thewilsonator) [reply spec/cpp_interface.dd:189]: Ah, one of them was a tab.

File: spec/cpp_interface.dd [hunk 9]
```diff
@@ -178,6 +178,88 @@ void main()
...
+$(P Then you can import the package containing the extern C++ declarations as follows:)
+
+---
+import ns;
+static assert(foo() == 1 && bar() == 2);
+---
+
+$(P Note that the `extern (C++, `ns`)` linkage attribute affects only the ABI (name mangling and calling convention) of
+  these declarations. Importing them follows the usual
+  $(LINK2 attribute.html#namespace, D module import semantics).)
```
Comment (PetarKirov) [inline spec/cpp_interface.dd:225]: Sorry, before I gave you suggestion with a mistake in it:

```suggestion
  $(LINK2 spec/module, D module import semantics).)
```

**Pull request merged**